jupyterhub/handlers/pages.py

```diff
@@ -24,6 +27,11 @@ class RootHandler(BaseHandler):
     Otherwise, renders login page.
     """
     def get(self):
+        next_url = self.get_argument('next', '')
+        if next_url and next_url.startswith('/user/'):
```

Comment:
This seems to work:
```
    def get(self):
        next_url = self.get_argument('next', '')
        if next_url and next_url.startswith(url_path_join(self.base_url, 'user') + '/'):
            # Redirect /hub/?next=... requests from single-user servers.
            url = url_path_join(self.hub.server.base_url, next_url[len(self.base_url):].lstrip('/'))
            self.redirect(url)
```

So on the second last line I strip the base URL from the next URL before tacking it onto the hub base URL.